examples/probability.dx

```diff
@@ -0,0 +1,297 @@
+  Rain (R), Sprinkler (S), and Wet grass (G).
+
+def wet ((rain_var, sprinkler_var, wet_var):
+            (DVar Rain & DVar Sprinkler & DVar Binary)) :
+        Probs Binary =
+    -- Rain is drawn from a bernoulli
+    rain  = sample rain_var $ bernoulli 0.2
+
+    -- Sprinkler is drawn conditional on rain. It is non-Markovian.
+    sprinkler_rain = cond_sample sprinkler_var [bernoulli 0.4, bernoulli 0.01] rain
+
+    -- CPT for producing wet grass conditioned on other variables
+    trans = for (s, r). [[bernoulli 0.0, bernoulli 0.8],
+                         [bernoulli 0.9, bernoulli 0.99]].s.r
+                         
+    -- Grass is draw conditional on Rain / Sprinkler
+    sample wet_var $ trans <| sprinkler_rain
+
+
+' Case where all the variables are True. $P(G=T, S=T, R=T)$
+
+show $ wet ({|obs=true|}, {|obs=true|}, {|obs=true|})
+
+(rain, sprinkle) = marginalize (\ (v2, v3). wet (lat v2, lat v3, var) ?? [False, True])
+
+
+' Marginal probability of rain given wet grass $P(R=T | G=T)$
+
+show $ AsProbsTable rain
+
+
+' This shows that we can sample from the distribution using the same code.
+
+[k1, k2] = splitKey $ newKey 0
+
+wet ({|sample=(one, k1)|}, {|sample=(one,k2)|}, var) 
+
+
+
+' ## Example 2: Dice
+
+' Dice values as represented by Fin
+
+Faces = Fin 6
+def diceval (x: Faces) : Int = (ordinal x) + 1
+
+
+def dice ((dice_vars1, dice_vars2, sum_var): (DVar Faces & DVar Faces & DVar (Fin 13))) :
+      Probs (Fin 13) =
+    dice1  = sample dice_vars1 $ uniform
+    dice2  = sample dice_vars2 $ uniform
+
+    -- Sum of dice rolls as a CPT table
+    sumdice = for (i, j). delta (((diceval i) + (diceval j))@_)
+
+    -- Dice are independent
+    sample sum_var $ sumdice <| (dice1 ,, dice2)
+
+' Sum of two dice
+
+show $ dice (var, var, var)
+
+-- Sum of two dice with one fixed.
+
+dice (var, {|obs= (3@_)|}, var)
+
+
+-- Dice values conditioned on summing to 3
+marginalize (\ (d1, d2). dice (lat d1, lat d2, {|obs=(3@_)|} ) ?? for i. True)
+
+' Sample through
+
+dice ({|sample= (one, k1)|}, {|sample=(one, k2)|}, var)
+
+
+
+' ## Example 3: HMM
+
+' Let's us easily generate arbitrary probability distributions
+
+instance [Arbitrary a] Arbitrary (Probs a)
+  arb = \k .
+      x = arb k
+      n = sum for i. abs x.i 
+      AsProbsTable for i. (abs x.i) / n
+
+
+' Very concise implementation of an HMM
+
+Z = Fin 5
+X = Fin 10
+
+
+def hmm ((hidden_vars, init_var): (m => DVar Z & DVar Z)) (x: m => X)
+        (transition : CDist Z Z) (emission: CDist X Z)
+    : Probs Z =
+
+    -- Sample an initial state
+    yieldState (sample init_var uniform) \zref .
+       for i.
+           -- Sample next state
+           z' = sample hidden_vars.i $ transition <| (get zref)
+
+           -- Factor in evidence
+           zref := obs x.i z' emission    
+
+' Compute an HMM over a simple sequence
+
+show $ hmm (for i:(Fin 5). var, var) (for i. (0@_)) (arb k1) (arb k2)
+
+' Do marginalization using a autograd (Currently doesn't work yet)
```

Comment:
What's the problem with that? Can you please open an issue?

Reply:
Will do. When there are datatypes in my auto-differentiation path (not the variables I am differentiating with respect to), I often get run-time errors. I will make some sample cases.